new version

marcelloceschia opened the pull request:
@dkgroot Kannst Du bitte mal darüber schauen, das mit den softkeys ist mir nicht klar

Commit:
new version

Commit:
Merge remote-tracking branch 'origin/master' into feature/refactored

Commit:
remove wrong execute

Commit:
fix missing argument

Commit:
fix missing host header value

Commit:
fix missing host header value

dkgroot commented:
Ich gucken gerade, aber bei mir geht irgendwas nicht (mit beide scripts). Konnte sein weil meine telefone durchgeschleift sind :-)
Es seiht aber gut aus. Die buttons reagieren correct (Auser Exit, der lauft nicht ganz).

Welchen teil mbz der softkeys ist nicht deutlich ?

Der neue version sieht viel professioneller aus, BTW !

marcelloceschia commented:
@dkgroot was soll denn exit machen? ein Stopp für alle devices?

dkgroot commented:
Es muss einen STOP des sendens und gleich zeiting der seite schliessen
das geht uber:
- App:Close (java) 
- Init:Services (alt)). 
- Softkey:Exit (macht meisten eher Previous Page)

Du Konntest denn stop and der URLDown action und der App:Close an der URL hangen wenn gewollt.

Commit:
use Init:Services on stop

dkgroot commented:
Found the cause for Device->push not working.
The contrent lenght needs to include the "XML=" part of the string as well

So instead of:
```
$post .= "Content-Length: ".strlen($xml)."\r\n\r\n";
$post .= "XML=" . $xml;
````
You need
```
$xml = "XML=".urlencode($xml);
Content-Length: ".strlen($xml)."\r\n\r\n
$post .= $xml;
```

Dies hatt dafur gesorgt das meine devices nie angefangen hatten zu zu horen.

marcelloceschia commented:
@dkgroot Das mit Init:Services habe ich verstanden, aber wer bekommt welche SoftKeys?

`if (!$isTalking) {` ist für die devices die nicht den ptt gestartet haben? Aber was soll dann lock und unlock machen?

Commit:
fixes Content-Length issue, thank you @dkgroot

dkgroot commented:
Lock start continues transmission (mann muss denn talk button nicht drucken zu sprechen) 'isTalking=true'
Unlock stop continues transmission (isTalking=false)

Es ist nur eine erweiterung, aber nicht wichtig.

marcelloceschia commented:
@dkgroot jetzt bin ich total verwirrt. Wenn ich das dann richtig lese, kann jeder der Exit macht alles beenden? 
Demzufolge wäre `action=mute` bzw  `action=unmute` korrekt?

dkgroot commented:
Damit hast du recht. Am moment zeigst du der page aber nur der anfanger der ptt.. Eigentlich sollte der page an alle mitglieder gepush werden, damit die auch 'talk' drucken konne.

BTW: ich habe das script nicht geschrieben, nur fur den Jozef auf den wiki geposted. Also lock/unlock is eher seine erfindung.

Du brauchst einen 'initiator' URI variable. Damit kannst du dann der initiele push machen zu denn anderen mitglieder. In der gleiche push kannst auch direct RTPMRx machen, damit die zuhoren.

Die mitglieder konnen dann auf der talk button drucken wenn sie sprechen wollen. Hatt ein mitglied aber lock gedruckt, dann konnen die andere nicht sprechen (Page Update musste dafur gepusht werden).

Exit sollte nur der eine teilnehmen rausnehmen. Ausser wenn das de initiator ist.

marcelloceschia commented:
Gut, dann passe ich das morgen mal an.
Werde die actions dann mal noch eindeutiger bezeichnen und noch einen parameter für den Initiator einfügen.

Ich weiß das du es nicht geschrieben hast, daher ist es aber noch wichtiger das die actions eindeutig sind 😉

dkgroot commented:
Ganz einverstanden !
Deine version wird ganz schon :-)

BTW: Mit `error_log("string", 0)` kannst du einfach debuggen uber syslog. z.b. in push($xml)
```
                error_log("pushed:" . $xml, 0);
                error_log("to:" . $this->ip, 0);
                error_log("response:" . $response, 0);
```

BTW: wir sollten der push response auswerten, damit mann weis ob es geklapt hatt (Und wer also mitmacht, konnte sogewohlt auch der initiator zeigen).

dkgroot commented on ptt.php:
```diff
@@ -1,156 +1,179 @@
+			}
+
+		}
+
```
error_log("pushed:" . $xml, 0);
                error_log("to:" . $this->ip, 0);
                error_log("response:" . $response, 0);

                // we should check the return value !

dkgroot commented on ptt.php:
```diff
@@ -1,156 +1,179 @@
+	const URI_START			= 'RTPMRx:%s:%d';
+
+	public $multicastAddress= '225.3.15.13';
+	public $multicastPort 	= 20480;
```
Comment required !
```
// The MulticastAddress has to between 225.0.0.0 to 239.255.255.255.
// Phones only accept  and even port number
```

dkgroot commented on ptt.php:
```diff
@@ -1,156 +1,179 @@
+$push2Talk->setAuthentication('cisco', 'cisco');
+$push2Talk->addDevices(array('10.0.2.225', '10.0.2.227'));
+
+// optional settings
```
Comment required !
```
// The MulticastAddress has to between 225.0.0.0 to 239.255.255.255.
// Phones only accept  and even port number
```

dkgroot commented on ptt.php:
```diff
@@ -1,156 +1,179 @@
+	$deviceName = $_GET['name'];
+
+	switch($action){
+		default:
```
We should have an initial action which:
- Pushed the page to all the other participants
- After pushing the page, it can perform the push2talk->start() action (Should on start once, not every page refresh)

dkgroot commented on ptt.php:
```diff
@@ -1,156 +1,179 @@
+		$response .=  "<URLDown>RTPMRx:Stop</URLDown>";
+		$response .=  "</SoftKeyItem>";
+
+		$response .=  "<SoftKeyItem>";
```
Lock indicates that one party is continuesly talking. The other participant should get a page refresh and they should on see an exit button. They are locked out, so to speak.
OneWay Audio

dkgroot commented on ptt.php:
```diff
@@ -1,156 +1,179 @@
+		$response .=  "<URLDown>RTPMTx:{$push2Talk->multicastAddress}:{$push2Talk->multicastPort}</URLDown>";
+		$response .=  "</SoftKeyItem>";
+
+		$response .=  "<SoftKeyItem>";
```
initiator:Exit -> stop and close page on all participants
participant:Exit -> close page only, take device out of devicesArray, to prevent further updates.

zipiju commented:
Hello guys, was updating the original PTT script and found out that #DEVICENAME# is not handled correctly in Softkeys on all the phones - the 7921 is throwing an error, 7965 on the other hand is okay with that. I assume the placeholder on 7921 is not replaced by MAC thus webserver is responding with something incomplete, maybe even with some error if phone is sending # in URL.
It will be safer not to use #DEVICENAME# in Softkeys and just use $MAC there.
In page push (and pull when used in SEP.cfg.xml) the #DEVICENAME# is working ok on those two platforms.

marcelloceschia committed:
change back to $deviceName

as requested by @zipiju

dkgroot committed:
Parse push response. Added syslog debugging (to ease development)

zipiju committed:
Prevent Softkeys sending the query string twice

This fixes Softkeys sending HTTP requests with duplicate query string
(like ptt.php?name=SEP002414xxxxxx?name=SEP002414xxxxxx&action=lock)

dkgroot committed:
Added originator and only allow close for origintor. Other participants can leave.

dkgroot commented:
@zipiju : Thanks for your help :+1:

dkgroot committed:
Added cookie handling for participants.
Fixed priority in execute closure (needs use ($priority) to access parent scope variable from closure).
Handle leave and stop seperately.

zipiju committed:
Small correcton

The MAC of the phone that has been commanded to pull PTT page is not
known at that time thus PTT URL has to contain MAC placeholder.

zipiju committed:
Make sure close softkey is not pushed to participants

zipiju commented:
Np. Was also taking a look at the cookies, but server here is complaining about headers already being sent. It however might be the output buffering settings.
Also not sure if you can actually share a cookie between the clients. With sessions you can pass SID and resume the session to modify the content on server. With cookie if the participant will leave I'm not sure he will be able to update the cookie on the other phones (if that was the plan to store who actually joined or left already).
I'm quite done for today, will take a nap and check/think about it later :).

dkgroot commented:
If they press the exit button, we still have the option of sending an event. On the Java phones we could use app:close event. That only leaves the old phones.

I didn't want to introduce any other dependencies (like DB etc), so it took the simple way of informing all the clients of the participants, but that will not hold up I think. What we could do is have the originator use a session to manage the participants, when a participant leaves he informs the originator using another push. Then the originator updates the session and pushes an update to all recipients.... It starting to become a little ugly again.

Have a nice rest !

zipiju commented:
The phone leaving can update the session itself if it will pass the SID to the server - "action=leave&SID=...". The session would be created by the phone starting the session - session_start() and the other phones will be told SID when asked to load PTT page. Then there can be a soft button like "PartList" (simmilar to ConfList) which can load a page that will list the participants stored in the session.
To inform the other phones about someone leaving or joning maybe we can use Play event to play a beep like in the conference.

dkgroot commented:
@zipiju: Ok that sounds like a nice solution ! Did you make these changes already on you local system ?

Also, marcello doesn't make it to complex
https://github.com/chan-sccp/ptt/commit/93f61fededcf3f3f41337c58315447fb325541a8#commitcomment-21100719

Might be better if i let you guys sort it out (If that's ok with you guys)

zipiju commented:
@dkgroot Not yet but will work on it as soon as possible. I'm okay with it.

zipiju commented:
Have the sessions working. Participants list and participants leaving still WIP. I'll be pushing modified script probably saturday if time allows.

zipiju committed:
Use sessions to store participants joined

Reworked to use sessions to store list of participants who really
joined.
Added participant list.
Crude version, updates to follow.
